test_conformance/api/test_queries.cpp

```diff
@@ -476,30 +476,124 @@ REGISTER_TEST(get_command_queue_info_compatibility)
-              (int)sizeof(cl_context_properties), (int)size );
+    log_error("ERROR: Returned size of context props is not valid! (expected 0 "
+              "or %d, got %d)\n",
+              (int)sizeof(cl_context_properties), (int)size);
     return -1;
 }
 
+REGISTER_TEST(get_context_info_mult_devices)
+{
+    // query multi-device context and perform objects comparability test
+    cl_uint maxComputeUnits = 0;
+    cl_int err =
+        clGetDeviceInfo(device, CL_DEVICE_MAX_COMPUTE_UNITS,
+                        sizeof(maxComputeUnits), &maxComputeUnits, NULL);
+    test_error_fail(err, "Unable to get maximal number of compute units");
+
+    cl_device_partition_property partitionProp[] = {
+        CL_DEVICE_PARTITION_EQUALLY,
+        static_cast<cl_device_partition_property>(maxComputeUnits / 2), 0
+    };
+
+    cl_uint num_devices = 0;
+    // how many sub-devices can we create?
+    err = clCreateSubDevices(device, partitionProp, 0, nullptr, &num_devices);
+    if (err != CL_SUCCESS || num_devices < 2)
+    {
+        log_info("Can't partition device, test not supported\n");
+        return TEST_SKIPPED_ITSELF;
+    }
+
+    // get the list of subDevices
+    SubDevicesScopeGuarded scope_guard(num_devices);
+    err = clCreateSubDevices(device, partitionProp, num_devices,
+                             scope_guard.sub_devices.data(), &num_devices);
+    if (err != CL_SUCCESS)
+    {
+        log_info("Can't partition device, test not supported\n");
+        return TEST_SKIPPED_ITSELF;
+    }
+
+    /* Create a multi device context */
+    clContextWrapper multi_device_context =
+        clCreateContext(NULL, (cl_uint)num_devices,
+                        scope_guard.sub_devices.data(), nullptr, nullptr, &err);
+    test_error_ret(err, "Unable to create testing context", TEST_FAIL);
+
+    err = clGetContextInfo(multi_device_context, CL_CONTEXT_NUM_DEVICES,
+                           sizeof(cl_uint), &num_devices, nullptr);
+    test_error_ret(err, "clGetContextInfo CL_CONTEXT_NUM_DEVICES failed\n",
+                   TEST_FAIL);
+
+    test_assert_error(num_devices == scope_guard.sub_devices.size(),
+                      "Context must contain exact number of devices\n");
+
+    std::vector<cl_device_id> devices(num_devices);
+    err = clGetContextInfo(multi_device_context, CL_CONTEXT_DEVICES,
+                           num_devices * sizeof(cl_device_id), devices.data(),
+                           nullptr);
+    test_error_ret(err, "clGetContextInfo CL_CONTEXT_DEVICES failed\n",
+                   TEST_FAIL);
+
+    test_assert_error(devices.size() == scope_guard.sub_devices.size(),
+                      "Size of devices arrays must be in sync\n");
+
+    for (cl_uint i = 0; i < devices.size(); i++)
+    {
+        test_error_ret(scope_guard.sub_devices[i] == devices[i],
+                       "Unexpected result returned by CL_CONTEXT_DEVICES query",
+                       TEST_FAIL);
```

Comment:
Observation: This requires the devices returned by `CL_CONTEXT_DEVICES` to be returned in the same order as the devices passed to `clCreateContext`.  This is likely the behavior most users will expect, but I don't think we require this in the spec.  Should we clarify `CL_CONTEXT_DEVICES` to document this requirement?

Reply:
Done!